### DISCUSSION TITLE
How to send str json format ping

### DISCUSSION THREAD

**Discussion by river-walras** [tarasko/picows#21] [Q&A] [answered]
It seems bybit requires user to send custom ping in `str` type not `byte` type. 

<img width="973" alt="image" src="https://github.com/user-attachments/assets/7854f4e1-5db5-4912-8990-b79b97f3e369">

Its offical [sdk](https://github.com/bybit-exchange/pybit/blob/master/pybit/_websocket_stream.py) using `self.custom_ping_message = json.dumps({"op": "ping"})` which returns a str type. However, I am using `orjson.dumps({"op": "ping"}` which returns byte type and the websocket will return `b'\x03\xe9'`. How to solve it?

```python
class Listener(WSListener):
    def __init__(self, logger, specific_ping_msg: bytes = None):
        self._log = logger
        self.msg_queue = asyncio.Queue()
        self._specific_ping_msg = specific_ping_msg

    def send_user_specific_ping(self, transport: WSTransport):
        if self._specific_ping_msg:
            transport.send(WSMsgType.TEXT, self._specific_ping_msg)
        else:
            transport.send_ping()

    def on_ws_connected(self, transport: WSTransport):
        self._log.info("Connected to Websocket...")

    def on_ws_disconnected(self, transport: WSTransport):
        self._log.info("Disconnected from Websocket.")

    def on_ws_frame(self, transport: WSTransport, frame: WSFrame):
        try:
            if frame.msg_type == WSMsgType.PING:
                transport.send_pong(
                    frame.get_payload_as_bytes()
                )  # if enabled auto pong, we don't need to send pong manually
                return
            # msg = orjson.loads(frame.get_payload_as_bytes())
            self.msg_queue.put_nowait(
                frame.get_payload_as_bytes()
            )  # I think we should decode the frame in the handler
        except Exception as e:
            self._log.error(f"Error processing message: {str(e)}")
   ```
However, If I send `json.dumps({"op": "ping"})`, I got:
```
Traceback (most recent call last):
  File "picows/picows.pyx", line 1080, in picows.picows.WSProtocol._auto_ping_loop
  File "picows/picows.pyx", line 344, in picows.picows.WSListener.send_user_specific_ping
  File "/root/tradebot-pro/tradebot/base.py", line 371, in send_user_specific_ping
    transport.send(WSMsgType.TEXT, self._specific_ping_msg)
  File "picows/picows.pyx", line 453, in picows.picows.WSTransport.send
  File "picows/picows.pyx", line 473, in picows.picows.WSTransport.send
  File "picows/picows.pyx", line 154, in picows.picows._unpack_bytes_like
TypeError: a bytes-like object is required, not 'str'
```

**Comment by river-walras** [comment]
> It seems bybit requires user to send custom ping in `str` type not `byte` type.
> 
> <img alt="image" width="973" src="https://private-user-images.githubusercontent.com/164357399/383962084-7854f4e1-5db5-4912-8990-b79b97f3e369.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA5ODM4MjgsIm5iZiI6MTczMDk4MzUyOCwicGF0aCI6Ii8xNjQzNTczOTkvMzgzOTYyMDg0LTc4NTRmNGUxLTVkYjUtNDkxMi04OTkwLWI3OWI5N2YzZTM2OS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwN1QxMjQ1MjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MzM2MDI0ZDRiZjgwMTkzNzEyMzViNzliODE3Y2NkMzdhYmRkYzJmOTk1NDM3ZDE2YjJiNjJiNWFhMjRlNzYyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.EXWyusRq5qaUJ5cbZ8Vcvarpsdlr7IU4Lx3ucudIlIA">
> Its offical [sdk](https://github.com/bybit-exchange/pybit/blob/master/pybit/_websocket_stream.py) using `self.custom_ping_message = json.dumps({"op": "ping"})` which returns a str type. However, I am using `orjson.dumps({"op": "ping"}` which returns byte type and the websocket will return `b'\x03\xe9'`. How to solve it?
> 
> ```python
> class Listener(WSListener):
>     def __init__(self, logger, specific_ping_msg: bytes = None):
>         self._log = logger
>         self.msg_queue = asyncio.Queue()
>         self._specific_ping_msg = specific_ping_msg
> 
>     def send_user_specific_ping(self, transport: WSTransport):
>         if self._specific_ping_msg:
>             transport.send(WSMsgType.TEXT, self._specific_ping_msg)
>         else:
>             transport.send_ping()
> 
>     def on_ws_connected(self, transport: WSTransport):
>         self._log.info("Connected to Websocket...")
> 
>     def on_ws_disconnected(self, transport: WSTransport):
>         self._log.info("Disconnected from Websocket.")
> 
>     def on_ws_frame(self, transport: WSTransport, frame: WSFrame):
>         try:
>             if frame.msg_type == WSMsgType.PING:
>                 transport.send_pong(
>                     frame.get_payload_as_bytes()
>                 )  # if enabled auto pong, we don't need to send pong manually
>                 return
>             # msg = orjson.loads(frame.get_payload_as_bytes())
>             self.msg_queue.put_nowait(
>                 frame.get_payload_as_bytes()
>             )  # I think we should decode the frame in the handler
>         except Exception as e:
>             self._log.error(f"Error processing message: {str(e)}")
> ```
> 
> However, If I send `json.dumps({"op": "ping"})`, I got:
> 
> ```
> Traceback (most recent call last):
>   File "picows/picows.pyx", line 1080, in picows.picows.WSProtocol._auto_ping_loop
>   File "picows/picows.pyx", line 344, in picows.picows.WSListener.send_user_specific_ping
>   File "/root/tradebot-pro/tradebot/base.py", line 371, in send_user_specific_ping
>     transport.send(WSMsgType.TEXT, self._specific_ping_msg)
>   File "picows/picows.pyx", line 453, in picows.picows.WSTransport.send
>   File "picows/picows.pyx", line 473, in picows.picows.WSTransport.send
>   File "picows/picows.pyx", line 154, in picows.picows._unpack_bytes_like
> TypeError: a bytes-like object is required, not 'str'
> ```

It is also wield, I receiving success `ping` message when sending byte type ping.
<img width="1487" alt="image" src="https://github.com/user-attachments/assets/3f3c2ff8-0f1d-4d0a-942e-142dc86b8591">
<img width="974" alt="image" src="https://github.com/user-attachments/assets/78e11fcc-15d8-4a71-9850-686d9ac1aa9f">
But still receiving the `b'\x03\xe9'` and one successful `ping` return corresponds to the error message `b'\x03\xe9'`. I have checked the frame type of the `b'\x03\xe9'` which is `<WSMsgType.CLOSE: 8>`, why would I receive a close frame?

**Comment by tarasko** [comment, marked as answer]
It seems that picows library initiating disconnect because you didn't let picows know that you have received user specific pong.

If you redefine ping messages by overriding WSListener.send_user_specific_ping, you should also redefine expected pong replies by redefining either WSListener.is_user_specific_pong or by calling transport.notify_user_specific_pong_received() from on_ws_frame.

Check the guide on auto ping feature. It explains that.
https://picows.readthedocs.io/en/stable/guides.html#auto-ping

Also picows logs that it has not received user specific pong within required time and that's why it has initiated disconnect. You should be able to see that in the log

**Reply by river-walras** [reply]
> It seems that picows library initiating disconnect because you didn't let picows know that you have received user specific pong.
> 
> If you redefine ping messages by overriding WSListener.send_user_specific_ping, you should also redefine expected pong replies by redefining either WSListener.is_user_specific_pong or by calling transport.notify_user_specific_pong_received() from on_ws_frame.
> 
> Check the guide on auto ping feature. It explains that. https://picows.readthedocs.io/en/stable/guides.html#auto-ping
> 
> Also picows logs that it has not received user specific pong within required time and that's why it has initiated disconnect. You should be able to see that in the log

I have tried to modify the `WSListener.is_user_specific_pong` by
```python
def WSListener.is_user_specific_pong():
    return False
```
But the problem is not fixed.

**Reply by river-walras** [reply]
> > It seems that picows library initiating disconnect because you didn't let picows know that you have received user specific pong.
> > If you redefine ping messages by overriding WSListener.send_user_specific_ping, you should also redefine expected pong replies by redefining either WSListener.is_user_specific_pong or by calling transport.notify_user_specific_pong_received() from on_ws_frame.
> > Check the guide on auto ping feature. It explains that. https://picows.readthedocs.io/en/stable/guides.html#auto-ping
> > Also picows logs that it has not received user specific pong within required time and that's why it has initiated disconnect. You should be able to see that in the log
> 
> I have tried to modify the `WSListener.is_user_specific_pong` by
> 
> ```python
> def WSListener.is_user_specific_pong():
>     return False
> ```
> 
> But the problem is not fixed.

I understand ! If I return False, picows think I did not received the reply. So I should handle them by calling transport.notify_user_specific_pong_received()
